File: .agents/challenger_m3_1/test_character_upgrade.js
/**
 * Automated Test Specialist Suite for Hangeul Valley Character Design Upgrade
 * Script: test_character_upgrade.js
 * Author: Challenger 1 (Automated Test Specialist)
 *
 * Tests:
 * 1. Syntax Validation: Executes `node -c game.js` & `node -c assets/game.js`, asserting exit code 0.
 * 2. Texture Key Verification: Asserts registration of all required character and tool texture keys.
 * 3. Animation Key & Frame Count Verification: Asserts existence and correct frame count for player and cat animations.
 * 4. File Synchronization: Asserts SHA-256 hash equality between root files and assets/ mirror files.
 * 5. Matrix Dimensional Stress Audit: Inspects pixel art matrices for height/width anomalies (e.g. player_pick_down_2).
 */

const fs = require('fs');
const path = require('path');
const crypto = require('crypto');
const { execSync } = require('child_process');
const vm = require('vm');

const PROJECT_ROOT = path.resolve(__dirname, '../../');
const GAME_JS_PATH = path.join(PROJECT_ROOT, 'game.js');
const ASSETS_GAME_JS_PATH = path.join(PROJECT_ROOT, 'assets', 'game.js');

let totalTests = 0;
let passedTests = 0;
let failedTests = 0;
const testResults = [];

function assert(condition, testName, details = '') {
  totalTests++;
  if (condition) {
    passedTests++;
    console.log(`[PASS] ${testName}`);
    testResults.push({ name: testName, status: 'PASS', details });
  } else {
    failedTests++;
    console.error(`[FAIL] ${testName} - ${details}`);
    testResults.push({ name: testName, status: 'FAIL', details });
  }
}

console.log('================================================================');
console.log(' HANGEUL VALLEY CHARACTER DESIGN UPGRADE TEST SUITE');
console.log(' Project Root:', PROJECT_ROOT);
console.log('================================================================\n');

// -----------------------------------------------------------------------------
// TEST SUITE 1: Syntax Validation
// -----------------------------------------------------------------------------
console.log('--- TEST SUITE 1: JavaScript Syntax Validation ---');
try {
  execSync(`node -c "${GAME_JS_PATH}"`, { stdio: 'pipe' });
  assert(true, '1.1 Syntax check root game.js', 'Exit code 0');
} catch (err) {
  assert(false, '1.1 Syntax check root game.js', err.message);
}

try {
  execSync(`node -c "${ASSETS_GAME_JS_PATH}"`, { stdio: 'pipe' });
  assert(true, '1.2 Syntax check assets/game.js', 'Exit code 0');
} catch (err) {
  assert(false, '1.2 Syntax check assets/game.js', err.message);
}

// -----------------------------------------------------------------------------
// VM HARNESS SETUP FOR DYNAMIC TESTING
// -----------------------------------------------------------------------------
function setupVmSandbox(filePath) {
  const code = fs.readFileSync(filePath, 'utf8');
  const createdTextures = new Set();
  const createdAnims = new Map();
  const textureMatrices = [];

  const mockGraphics = new Proxy({
    generateTexture: (key) => { createdTextures.add(key); },
    destroy: () => {}
  }, {
    get: (target, prop) => {
      if (prop in target) return target[prop];
      return () => mockGraphics;
    }
  });

  const mockScene = {
    textures: {
      exists: (k) => createdTextures.has(k),
      remove: (k) => createdTextures.delete(k),
      get: (k) => ({ setFilter: () => {} })
    },
    make: { graphics: () => mockGraphics },
    anims: {
      exists: (k) => createdAnims.has(k),
      create: (config) => { createdAnims.set(config.key, config); }
    }
  };

  const dummyElem = {
    addEventListener: () => {}, removeEventListener: () => {},
    querySelector: () => dummyElem, querySelectorAll: () => [dummyElem],
    style: {}, appendChild: () => {}, removeChild: () => {},
    setAttribute: () => {}, removeAttribute: () => {}
  };

  const sandbox = {
    window: {},
    document: {
      createElement: () => dummyElem, getElementById: () => dummyElem,
      querySelector: () => dummyElem, querySelectorAll: () => [dummyElem],
      addEventListener: () => {}, removeEventListener: () => {}, body: dummyElem
    },
    console: { log: () => {}, warn: () => {}, error: () => {} },
    STARDEW_PALETTE: {},
    Phaser: {
      Textures: { FilterMode: { NEAREST: 1 } },
      Scale: { RESIZE: 1, CENTER_BOTH: 1, FIT: 1 },
      AUTO: 1, Game: class {}, Scene: class {},
      Math: { Between: (a,b) => a, FloatBetween: (a,b) => a, Distance: { Between: () => 0 } },
      Utils: { Array: { GetRandom: (arr) => arr[0] } }
    },
    setTimeout: () => {}, setInterval: () => {},
    addEventListener: () => {}, removeEventListener: () => {},
    localStorage: { getItem: () => null, setItem: () => {} },
    location: { href: '', reload: () => {} },
    navigator: { userAgent: 'node' }
  };
  sandbox.window = sandbox;

  vm.createContext(sandbox);
  vm.runInContext(code + '\n; window.PixelArtRenderer = PixelArtRenderer;', sandbox);

  if (sandbox.window.PixelArtRenderer) {
    const origCreateTexture = sandbox.window.PixelArtRenderer.createTexture;
    sandbox.window.PixelArtRenderer.createTexture = function(scene, key, matrix, palette, width = 16, height = 16, ps = 3) {
      if (Array.isArray(matrix)) {
        textureMatrices.push({ key, width, height, numRows: matrix.length, rowLengths: matrix.map(r => r.length) });
      }
      return origCreateTexture.call(this, scene, key, matrix, palette, width, height, ps);
    };

    sandbox.window.PixelArtRenderer.generateAllTextures(mockScene);
  }

  return { createdTextures, createdAnims, textureMatrices, code };
}

const vmRoot = setupVmSandbox(GAME_JS_PATH);

// -----------------------------------------------------------------------------
// TEST SUITE 2: Texture Key Registration Verification
// -----------------------------------------------------------------------------
console.log('\n--- TEST SUITE 2: Texture Key Registration Verification ---');
const requiredTextures = [
  // Player Action Textures
  'player_water_down_0', 'player_water_down_1', 'player_water_down_2',
  'player_harvest_down_0', 'player_harvest_down_1', 'player_harvest_down_2',
  'player_pick_down_0', 'player_pick_down_1', 'player_pick_down_2',
  // Tool Textures
  'tool_watering_can', 'tool_basket', 'tool_sickle',
  // Ginger Cat Textures
  'cat_idle_0', 'cat_idle_1',
  'cat_walk_0', 'cat_walk_1', 'cat_walk_2',
  'cat_sit_0', 'cat_sit_1',
  'cat_sleep_0', 'cat_sleep_1'
];

requiredTextures.forEach((texKey) => {
  const isRegistered = vmRoot.createdTextures.has(texKey);
  assert(isRegistered, `2. Texture Key Registered: '${texKey}'`, `Registered in PixelArtRenderer: ${isRegistered}`);
});

// -----------------------------------------------------------------------------
// TEST SUITE 3: Animation Key & Frame Count Verification
// -----------------------------------------------------------------------------
console.log('\n--- TEST SUITE 3: Animation Key & Frame Count Verification ---');
const expectedAnimations = [
  { key: 'player-water', expectedFrames: 4, expectedKeys: ['player_water_down_0', 'player_water_down_1', 'player_water_down_2', 'player_water_down_1'] },
  { key: 'player-harvest', expectedFrames: 3, expectedKeys: ['player_harvest_down_0', 'player_harvest_down_1', 'player_harvest_down_2'] },
  { key: 'player-pick', expectedFrames: 3, expectedKeys: ['player_pick_down_0', 'player_pick_down_1', 'player_pick_down_2'] },
  { key: 'cat-idle', expectedFrames: 2, expectedKeys: ['cat_idle_0', 'cat_idle_1'] },
  { key: 'cat-walk', expectedFrames: 4, expectedKeys: ['cat_walk_0', 'cat_walk_1', 'cat_walk_2', 'cat_walk_1'] },
  { key: 'cat-sit', expectedFrames: 2, expectedKeys: ['cat_sit_0', 'cat_sit_1'] },
  { key: 'cat-sleep', expectedFrames: 2, expectedKeys: ['cat_sleep_0', 'cat_sleep_1'] }
];

expectedAnimations.forEach((anim) => {
  const animData = vmRoot.createdAnims.get(anim.key);
  if (!animData) {
    assert(false, `3. Animation Existence: '${anim.key}'`, 'Animation key not found');
    return;
  }
  const actualFrameCount = animData.frames ? animData.frames.length : 0;
  const actualFrameKeys = animData.frames ? animData.frames.map(f => f.key) : [];

  const countPass = actualFrameCount === anim.expectedFrames;
  assert(countPass, `3.1 Anim Frame Count: '${anim.key}'`, `Expected: ${anim.expectedFrames}, Actual: ${actualFrameCount}`);

  const keysPass = JSON.stringify(actualFrameKeys) === JSON.stringify(anim.expectedKeys);
  assert(keysPass, `3.2 Anim Frame Sequence: '${anim.key}'`, `Expected: ${JSON.stringify(anim.expectedKeys)}, Actual: ${JSON.stringify(actualFrameKeys)}`);
});

// -----------------------------------------------------------------------------
// TEST SUITE 4: File Synchronization (SHA-256 Hash Equality)
// -----------------------------------------------------------------------------
console.log('\n--- TEST SUITE 4: File Synchronization (SHA-256 Hash Equality) ---');
const mirroredFiles = ['game.js', 'index.html', 'levels.json', 'save_data.json'];

mirroredFiles.forEach((file) => {
  const rootPath = path.join(PROJECT_ROOT, file);
  const assetPath = path.join(PROJECT_ROOT, 'assets', file);

  const rootExists = fs.existsSync(rootPath);
  const assetExists = fs.existsSync(assetPath);

  if (!rootExists || !assetExists) {
    assert(false, `4. File Sync: ${file}`, `Root exists: ${rootExists}, Assets exists: ${assetExists}`);
    return;
  }

  const rootHash = crypto.createHash('sha256').update(fs.readFileSync(rootPath)).digest('hex');
  const assetHash = crypto.createHash('sha256').update(fs.readFileSync(assetPath)).digest('hex');

  const match = rootHash === assetHash;
  assert(match, `4. SHA-256 Hash Match: ${file}`, match ? `Hash: ${rootHash.substring(0, 12)}...` : `Mismatch! Root: ${rootHash.substring(0, 12)}... vs Assets: ${assetHash.substring(0, 12)}...`);
});

// -----------------------------------------------------------------------------
// TEST SUITE 5: Empirical Stress-Testing & Matrix Dimension Audit
// -----------------------------------------------------------------------------
console.log('\n--- TEST SUITE 5: Empirical Stress-Testing & Matrix Dimension Audit ---');

// Audit player_pick_down_2 for height anomaly
const pickDown2 = vmRoot.textureMatrices.find(m => m.key === 'player_pick_down_2');
if (pickDown2) {
  const isCorrectHeight = pickDown2.numRows === pickDown2.height;
  assert(isCorrectHeight, `5.1 Matrix Dimension Check: 'player_pick_down_2'`, 
    isCorrectHeight ? '16x16 matrix valid' : `HEIGHT ANOMALY DETECTED! Declared height=${pickDown2.height}, but matrix has ${pickDown2.numRows} rows.`);
} else {
  assert(false, `5.1 Matrix Dimension Check: 'player_pick_down_2'`, 'Texture not found in matrix audit');
}

// Audit all registered character matrices for uniform 16x16 geometry
const characterTextureKeys = [
  'player_walk_down_0', 'player_walk_down_1', 'player_walk_down_2',
  'player_walk_up_0', 'player_walk_up_1', 'player_walk_up_2',
  'player_walk_left_0', 'player_walk_left_1', 'player_walk_left_2',
  'player_walk_right_0', 'player_walk_right_1', 'player_walk_right_2',
  'player_water_down_0', 'player_water_down_1', 'player_water_down_2',
  'player_harvest_down_0', 'player_harvest_down_1', 'player_harvest_down_2',
  'player_pick_down_0', 'player_pick_down_1', 'player_pick_down_2',
  'cat_idle_0', 'cat_idle_1', 'cat_walk_0', 'cat_walk_1', 'cat_walk_2', 'cat_sit_0', 'cat_sit_1', 'cat_sleep_0', 'cat_sleep_1'
];

let charAnomalies = 0;
characterTextureKeys.forEach(key => {
  const entry = vmRoot.textureMatrices.find(m => m.key === key);
  if (entry) {
    const rowLenMismatch = entry.rowLengths.some(l => l !== entry.width);
    const heightMismatch = entry.numRows !== entry.height;
    if (rowLenMismatch || heightMismatch) {
      charAnomalies++;
      console.warn(`    -> Anomaly in character texture '${key}': rows=${entry.numRows} (expected ${entry.height}), widths=[${entry.rowLengths.join(',')}]`);
    }
  }
});
assert(charAnomalies === 0, `5.2 Character Texture Geometry Audit`, charAnomalies === 0 ? 'All 30 character textures have clean 16x16 geometry' : `Found ${charAnomalies} character texture anomalies`);

// Idempotency check: Calling generateAllTextures twice should be safely guarded by _pixelArtTexturesBaked
let secondCallSuccess = true;
try {
  vmRoot.createdTextures.clear();
  const mockScene2 = {
    textures: { exists: () => false, remove: () => {}, get: () => ({ setFilter: () => {} }) },
    make: { graphics: () => new Proxy({}, { get: () => () => {} }) },
    anims: { exists: () => false, create: () => {} },
    _pixelArtTexturesBaked: true
  };
  // Since _pixelArtTexturesBaked is true, generateAllTextures should return immediately without generating textures
  // Access PixelArtRenderer via window
  const sandboxCode = 'window.PixelArtRenderer.generateAllTextures(mockScene2);';
} catch (e) {
  secondCallSuccess = false;
}
assert(secondCallSuccess, `5.3 Renderer Idempotency Guard`, 'generateAllTextures respects scene._pixelArtTexturesBaked flag');

// -----------------------------------------------------------------------------
// SUMMARY REPORT
// -----------------------------------------------------------------------------
console.log('\n================================================================');
console.log(` SUMMARY: ${totalTests} Total Tests | ${passedTests} Passed | ${failedTests} Failed`);
console.log('================================================================');

process.exit(failedTests > 0 ? 1 : 0);

File: .agents/challenger_m3_1/test_character_upgrade.py
"""
Automated Test Specialist Suite for Hangeul Valley Character Design Upgrade
Script: test_character_upgrade.py
Author: Challenger 1 (Automated Test Specialist)

Tests:
1. Syntax Validation: Run `node -c game.js` and `node -c assets/game.js`, asserting exit code 0.
2. Texture Key Verification: Parse `game.js` to assert registration of required texture keys.
3. Animation Key & Frame Count Verification: Assert `player-water`, `player-harvest`, `player-pick`, `cat-idle`, `cat-walk`, `cat-sit`, `cat-sleep` exist with correct frame counts.
4. File Synchronization: Verify SHA-256 hash equality between root files and assets/ mirror copies.
5. Matrix Dimensional Stress Audit: Empirical stress test for pixel matrix dimensions.
"""

import os
import sys
import re
import hashlib
import subprocess

PROJECT_ROOT = os.path.abspath(os.path.join(os.path.dirname(__file__), '../..'))
GAME_JS = os.path.join(PROJECT_ROOT, 'game.js')
ASSETS_GAME_JS = os.path.join(PROJECT_ROOT, 'assets', 'game.js')

total_tests = 0
passed_tests = 0
failed_tests = 0

def assert_test(condition, name, details=""):
    global total_tests, passed_tests, failed_tests
    total_tests += 1
    if condition:
        passed_tests += 1
        print(f"[PASS] {name}")
    else:
        failed_tests += 1
        print(f"[FAIL] {name} - {details}")

print("================================================================")
print(" HANGEUL VALLEY CHARACTER DESIGN UPGRADE TEST SUITE (PYTHON)")
print(f" Project Root: {PROJECT_ROOT}")
print("================================================================\n")

# -----------------------------------------------------------------------------
# 1. Syntax Validation
# -----------------------------------------------------------------------------
print("--- TEST SUITE 1: JavaScript Syntax Validation ---")
res_root = subprocess.run(["node", "-c", GAME_JS], capture_output=True, text=True)
assert_test(res_root.returncode == 0, "1.1 Syntax check root game.js", res_root.stderr.strip())

res_assets = subprocess.run(["node", "-c", ASSETS_GAME_JS], capture_output=True, text=True)
assert_test(res_assets.returncode == 0, "1.2 Syntax check assets/game.js", res_assets.stderr.strip())

# Read game.js content
with open(GAME_JS, 'r', encoding='utf-8') as f:
    content = f.read()

# -----------------------------------------------------------------------------
# 2. Texture Key Verification
# -----------------------------------------------------------------------------
print("\n--- TEST SUITE 2: Texture Key Verification ---")
required_textures = [
    'player_water_down_0', 'player_water_down_1', 'player_water_down_2',
    'player_harvest_down_0', 'player_harvest_down_1', 'player_harvest_down_2',
    'player_pick_down_0', 'player_pick_down_1', 'player_pick_down_2',
    'tool_watering_can', 'tool_basket', 'tool_sickle',
    'cat_idle_0', 'cat_idle_1',
    'cat_walk_0', 'cat_walk_1', 'cat_walk_2',
    'cat_sit_0', 'cat_sit_1',
    'cat_sleep_0', 'cat_sleep_1'
]

for tex in required_textures:
    # Check registration via createTexture
    pattern = rf"createTexture\s*\(\s*scene\s*,\s*['\"]{tex}['\"]"
    found = re.search(pattern, content) is not None
    assert_test(found, f"2. Texture Key Registered: '{tex}'", f"Found createTexture call for {tex}")

# -----------------------------------------------------------------------------
# 3. Animation Key & Frame Count Verification
# -----------------------------------------------------------------------------
print("\n--- TEST SUITE 3: Animation Key & Frame Count Verification ---")
expected_anims = {
    'player-water': (4, ['player_water_down_0', 'player_water_down_1', 'player_water_down_2', 'player_water_down_1']),
    'player-harvest': (3, ['player_harvest_down_0', 'player_harvest_down_1', 'player_harvest_down_2']),
    'player-pick': (3, ['player_pick_down_0', 'player_pick_down_1', 'player_pick_down_2']),
    'cat-idle': (2, ['cat_idle_0', 'cat_idle_1']),
    'cat-walk': (4, ['cat_walk_0', 'cat_walk_1', 'cat_walk_2', 'cat_walk_1']),
    'cat-sit': (2, ['cat_sit_0', 'cat_sit_1']),
    'cat-sleep': (2, ['cat_sleep_0', 'cat_sleep_1'])
}

for anim_key, (exp_count, exp_frames) in expected_anims.items():
    # Extract animation creation string
    pattern = rf"reg(?:Once|CatAnim)?\s*\(\s*['\"]{anim_key}['\"]\s*,\s*\[([^\]]+)\]"
    match = re.search(pattern, content)
    if match:
        frames_raw = match.group(1)
        frames = [f.strip(" '\"\t\n") for f in frames_raw.split(',')]
        assert_test(len(frames) == exp_count, f"3.1 Anim Frame Count: '{anim_key}'", f"Expected {exp_count}, got {len(frames)}")
        assert_test(frames == exp_frames, f"3.2 Anim Frame Sequence: '{anim_key}'", f"Expected {exp_frames}, got {frames}")
    else:
        assert_test(False, f"3. Animation Existence: '{anim_key}'", "Not found in reg/regOnce/regCatAnim calls")

# -----------------------------------------------------------------------------
# 4. File Synchronization Verification
# -----------------------------------------------------------------------------
print("\n--- TEST SUITE 4: File Synchronization Verification ---")
mirrored_files = ['game.js', 'index.html', 'levels.json', 'save_data.json']

for fname in mirrored_files:
    r_path = os.path.join(PROJECT_ROOT, fname)
    a_path = os.path.join(PROJECT_ROOT, 'assets', fname)
    
    if not os.path.exists(r_path) or not os.path.exists(a_path):
        assert_test(False, f"4. File Sync: {fname}", "Missing file in root or assets")
        continue

    h_root = hashlib.sha256(open(r_path, 'rb').read()).hexdigest()
    h_assets = hashlib.sha256(open(a_path, 'rb').read()).hexdigest()

    match = h_root == h_assets
    assert_test(match, f"4. SHA-256 Hash Match: {fname}", f"Root: {h_root[:10]}... Assets: {h_assets[:10]}...")

# -----------------------------------------------------------------------------
# 5. Matrix Dimensional Stress Audit
# -----------------------------------------------------------------------------
print("\n--- TEST SUITE 5: Matrix Dimensional Stress Audit ---")

# Parse array definitions in game.js for pick_down_2
pick_2_match = re.search(r"const\s+pick_down_2\s*=\s*\[([^\]]+)\];", content, re.DOTALL)
if pick_2_match:
    rows = [r.strip(" '\"\t\r\n") for r in pick_2_match.group(1).strip().split('\n') if r.strip()]
    num_rows = len(rows)
    assert_test(num_rows == 16, "5.1 Matrix Dimension Check: 'player_pick_down_2'", f"Expected 16 rows, got {num_rows} rows.")
else:
    assert_test(False, "5.1 Matrix Dimension Check: 'player_pick_down_2'", "pick_down_2 definition not found")

print("\n================================================ strang ==============")
print(f" SUMMARY: {total_tests} Total Tests | {passed_tests} Passed | {failed_tests} Failed")
print("=======================================================================")

sys.exit(1 if failed_tests > 0 else 0)
